File: dataset/ZeroShot/GEMINI_gemini-2.5-pro/Task154.java
import java.util.*;

public class Task154 {

    /**
     * Performs an iterative Depth-First Search (DFS) on a graph.
     *
     * @param V         The number of vertices in the graph.
     * @param adj       The adjacency list representation of the graph.
     * @param startNode The node to start the search from.
     * @return A list of integers representing the DFS traversal order.
     */
    public List<Integer> dfs(int V, List<List<Integer>> adj, int startNode) {
        // --- Security: Input Validation ---
        if (V <= 0 || startNode < 0 || startNode >= V || adj == null || adj.size() != V) {
            // Return an empty list for invalid input to prevent exceptions.
            return new ArrayList<>();
        }

        List<Integer> result = new ArrayList<>();
        boolean[] visited = new boolean[V];
        
        // Using ArrayDeque as a Stack is generally preferred over the legacy Stack class.
        Deque<Integer> stack = new ArrayDeque<>();

        stack.push(startNode);

        while (!stack.isEmpty()) {
            int u = stack.pop();

            if (!visited[u]) {
                visited[u] = true;
                result.add(u);

                // Get all adjacent vertices of the popped vertex u.
                // We iterate in reverse to approximate the order of a recursive DFS.
                List<Integer> neighbors = adj.get(u);
                if (neighbors != null) {
                    for (int i = neighbors.size() - 1; i >= 0; i--) {
                        int v = neighbors.get(i);
                        // --- Security: Check neighbor validity ---
                        if (v >= 0 && v < V && !visited[v]) {
                            stack.push(v);
                        }
                    }
                }
            }
        }
        return result;
    }
    
    // Helper function to add an edge to the graph
    private static void addEdge(List<List<Integer>> adj, int u, int v) {
        if (u >= 0 && u < adj.size() && v >= 0 && v < adj.size()) {
            adj.get(u).add(v);
        }
    }

    public static void main(String[] args) {
        Task154 solution = new Task154();
        
        // --- Test Cases ---
        System.out.println("--- 5 Test Cases for DFS ---");

        // Test Case 1: Connected Graph, Start 0
        int V1 = 7;
        List<List<Integer>> adj1 = new ArrayList<>(V1);
        for (int i = 0; i < V1; i++) adj1.add(new ArrayList<>());
        addEdge(adj1, 0, 1); addEdge(adj1, 0, 3);
        addEdge(adj1, 1, 2); addEdge(adj1, 1, 4);
        addEdge(adj1, 2, 1);
        addEdge(adj1, 3, 0); addEdge(adj1, 3, 4); addEdge(adj1, 3, 5);
        addEdge(adj1, 4, 1); addEdge(adj1, 4, 3);
        addEdge(adj1, 5, 3); addEdge(adj1, 5, 6);
        addEdge(adj1, 6, 5);
        System.out.println("Test Case 1 (Connected Graph, Start 0): " + solution.dfs(V1, adj1, 0));

        // Test Case 2: Connected Graph, Start 4
        System.out.println("Test Case 2 (Connected Graph, Start 4): " + solution.dfs(V1, adj1, 4));

        // Test Case 3: Disconnected Graph, Start 0
        int V3 = 5;
        List<List<Integer>> adj3 = new ArrayList<>(V3);
        for (int i = 0; i < V3; i++) adj3.add(new ArrayList<>());
        addEdge(adj3, 0, 1); addEdge(adj3, 1, 0); addEdge(adj3, 1, 2); addEdge(adj3, 2, 1); // Component 1
        addEdge(adj3, 3, 4); addEdge(adj3, 4, 3); // Component 2
        System.out.println("Test Case 3 (Disconnected Graph, Start 0): " + solution.dfs(V3, adj3, 0));

        // Test Case 4: Line Graph, Start 2
        int V4 = 5;
        List<List<Integer>> adj4 = new ArrayList<>(V4);
        for (int i = 0; i < V4; i++) adj4.add(new ArrayList<>());
        addEdge(adj4, 0, 1);
        addEdge(adj4, 1, 0); addEdge(adj4, 1, 2);
        addEdge(adj4, 2, 1); addEdge(adj4, 2, 3);
        addEdge(adj4, 3, 2); addEdge(adj4, 3, 4);
        addEdge(adj4, 4, 3);
        System.out.println("Test Case 4 (Line Graph, Start 2): " + solution.dfs(V4, adj4, 2));

        // Test Case 5: Single Node Graph, Start 0
        int V5 = 1;
        List<List<Integer>> adj5 = new ArrayList<>(V5);
        adj5.add(new ArrayList<>());
        System.out.println("Test Case 5 (Single Node Graph, Start 0): " + solution.dfs(V5, adj5, 0));
    }
}

File: dataset/Vanilla - Copy/GEMINI_gemini-2.5-pro/Task154.cpp
#include <iostream>
#include <vector>
#include <list>

class Graph {
private:
    int V; // Number of vertices
    std::vector<std::list<int>> adj; // Adjacency list

    /**
     * A recursive helper function for DFS.
     * @param v The current vertex.
     * @param visited A boolean vector to track visited vertices.
     * @param result The vector to store the DFS traversal order.
     */
    void DFSUtil(int v, std::vector<bool>& visited, std::vector<int>& result) {
        visited[v] = true;
        result.push_back(v);

        for (int neighbor : adj[v]) {
            if (!visited[neighbor]) {
                DFSUtil(neighbor, visited, result);
            }
        }
    }

public:
    /**
     * Graph constructor.
     * @param V Number of vertices in the graph.
     */
    Graph(int V) {
        this->V = V;
        adj.resize(V);
    }

    /**
     * Adds an edge to the graph.
     * @param v The source vertex.
     * @param w The destination vertex.
     */
    void addEdge(int v, int w) {
        if (v < V && w < V) {
            adj[v].push_back(w);
        }
    }

    /**
     * Performs Depth First Search traversal starting from a given node.
     * @param startNode The starting node for the DFS.
     * @return A vector of integers representing the DFS traversal.
     */
    std::vector<int> performDFS(int startNode) {
        std::vector<int> result;
        if (startNode >= V) {
            std::cout << "Start node is out of bounds." << std::endl;
            return result; 
        }
        std::vector<bool> visited(V, false);
        DFSUtil(startNode, visited, result);
        return result;
    }
};

void printVector(const std::vector<int>& vec) {
    for (int val : vec) {
        std::cout << val << " ";
    }
    std::cout << std::endl;
}

int main() {
    // Test Case 1: Simple linear graph
    std::cout << "Test Case 1:" << std::endl;
    Graph g1(4);
    g1.addEdge(0, 1);
    g1.addEdge(1, 2);
    g1.addEdge(2, 3);
    std::cout << "DFS starting from node 0: ";
    printVector(g1.performDFS(0));

    // Test Case 2: Graph with a fork
    std::cout << "\nTest Case 2:" << std::endl;
    Graph g2(5);
    g2.addEdge(0, 1);
    g2.addEdge(0, 2);
    g2.addEdge(1, 3);
    g2.addEdge(1, 4);
    std::cout << "DFS starting from node 0: ";
    printVector(g2.performDFS(0));

    // Test Case 3: Graph with a cycle
    std::cout << "\nTest Case 3:" << std::endl;
    Graph g3(4);
    g3.addEdge(0, 1);
    g3.addEdge(1, 2);
    g3.addEdge(2, 0);
    g3.addEdge(2, 3);
    std::cout << "DFS starting from node 0: ";
    printVector(g3.performDFS(0));

    // Test Case 4: Disconnected graph (DFS only visits connected component)
    std::cout << "\nTest Case 4:" << std::endl;
    Graph g4(5);
    g4.addEdge(0, 1);
    g4.addEdge(0, 2);
    g4.addEdge(3, 4);
    std::cout << "DFS starting from node 0: ";
    printVector(g4.performDFS(0));
    std::cout << "DFS starting from node 3: ";
    printVector(g4.performDFS(3));

    // Test Case 5: Single-node graph
    std::cout << "\nTest Case 5:" << std::endl;
    Graph g5(1);
    std::cout << "DFS starting from node 0: ";
    printVector(g5.performDFS(0));

    return 0;
}